### ISSUE TITLE
Lower Precision Grad Reduction

### ISSUE THREAD

**Issue by jfc4050** [jax-ml/jax#27496]
### Description

hello!

I am observing the following when using data parallelism and mixed precision
1. gradients computed in half precision in the backward pass
2. gradient reduction (inserted by partitioner) occurs in half precision
3. gradients then casted to full precision for optimizer step

whereas my expectation would be that the gradient reduction occurs after the cast (in full precision). On GPUs, Nvidia goes even further and [has wgrad kernels accumulate directly to a full precision buffer](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/examples/advanced_optimizations.html#Gradient-accumulation-fusion) (thus avoiding lossy casting sequences like fp32 GEMM accumulator -> bf16 output -> fp32 accumulator). How would you suggest I go about doing the same in JAX?

Please let me know if this issue should go somewhere else, I wasn't sure if it should be filed under XLA, JAX, or Flax.

Here's a minimal reproable example, and snapshots of the compiler HLO before and after the partitioner. Note: i did experiment with passing `jax.lax.Precision.HIGHEST`, and I see it reflected in the HLO, but it doesn't have any effect on the actual behavior.

```python
import os

os.environ["XLA_FLAGS"] = "--xla_dump_to=./dump_dir --xla_dump_hlo_pass_re=.*"


import flax.nnx
import jax.experimental.mesh_utils
import jax.numpy as jnp
import optax
from flax.nnx import Linear, Optimizer, Rngs
from jax import Array
from jax.sharding import Mesh, NamedSharding, PartitionSpec

model = Linear(
    in_features=8,
    out_features=16,
    use_bias=False,
    dtype=jnp.bfloat16,
    param_dtype=jnp.float32,
    rngs=Rngs(0),
)
optimizer = Optimizer(model, optax.adam(1e-3))


@flax.nnx.jit
def train_step(model: Linear, optimizer: Optimizer, x: Array, y_gt: Array) -> None:
    def loss_fn(model: Linear) -> Array:
        y_pred = model(x)
        return ((y_pred - y_gt) ** 2).mean()

    loss, grad = flax.nnx.value_and_grad(loss_fn)(model)

    optimizer.update(grad)


mesh = Mesh(
    devices=jax.experimental.mesh_utils.create_device_mesh(mesh_shape=(8,)),
    axis_names=("DP"),
)

with mesh:
    x = jnp.ones((8, 8), dtype=jnp.bfloat16, device=NamedSharding(mesh, PartitionSpec("DP", None)))
    y_gt = jnp.ones(
        (8, 16), dtype=jnp.bfloat16, device=NamedSharding(mesh, PartitionSpec("DP", None))
    )

    train_step(model, optimizer, x, y_gt)
```

if we look at the stableHLO we see the gradients are computed then upcasted
```
%9 = stablehlo.dot_general %8, %arg5, contracting_dims = [0] x [0], precision = [DEFAULT, DEFAULT] : (tensor<8x16xbf16>, tensor<8x8xbf16>) -> tensor<16x8xbf16>
%10 = stablehlo.transpose %9, dims = [1, 0] : (tensor<16x8xbf16>) -> tensor<8x16xbf16>
%11 = stablehlo.convert %10 : (tensor<8x16xbf16>) -> tensor<8x16xf32>
```

looking at the HLO right before partitioning - so far so good.
module_0029.jit_train_step.0009.spmd-partitioner.after_sharding-propagation.before_spmd-partitioning
```
dot = bf16[8,16]{0,1} dot(Arg_5.6, multiply.34), lhs_contracting_dims={0}, rhs_contracting_dims={0}, sharding={replicated}, metadata={op_name="jit(train_step)/jit(main)/transpose" source_file=".venv/lib/python3.12/site-packages/flax/nnx/nn/linear.py" source_line=367}
convert.37 = f32[8,16]{0,1} convert(dot), sharding={replicated}, metadata={op_name="jit(train_step)/jit(main)/convert_element_type" source_file=".venv/lib/python3.12/site-packages/flax/nnx/nn/dtypes.py" source_line=79}
```

after the partitioner runs. see that the all-reduce is inserted between the dot and the cast.
module_0029.jit_train_step.0010.spmd-cleanup.after_pipeline-start.before_dce
```
dot.2 = bf16[8,16]{1,0} dot(param.1, multiply.3), lhs_contracting_dims={0}, rhs_contracting_dims={0}, metadata={op_name="jit(train_step)/jit(main)/transpose" source_file=".venv/lib/python3.12/site-packages/flax/nnx/nn/linear.py" source_line=367}
all-reduce = bf16[8,16]{1,0} all-reduce(dot.2), channel_id=1, replica_groups=[1,8]<=[8], use_global_device_ids=true, to_apply=add.clone, sharding={replicated}, metadata={op_name="jit(train_step)/jit(main)/transpose" source_file=".venv/lib/python3.12/site-packages/flax/nnx/nn/linear.py" source_line=367}
convert.1 = f32[8,16]{0,1} convert(all-reduce), sharding={replicated}, metadata={op_name="jit(train_step)/jit(main)/convert_element_type" source_file=".venv/lib/python3.12/site-packages/flax/nnx/nn/dtypes.py" source_line=79}
```

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.36
jaxlib: 0.4.36
numpy:  2.1.3
python: 3.12.6 (main, Mar 23 2025, 12:21:25) [GCC 11.4.1 20230605 (Red Hat 11.4.1-2)]
device info: NVIDIA H100 80GB HBM3-8, 8 local devices"
process_count: 1
platform: uname_result(system='Linux', node='ip-172-31-37-73.ec2.internal', release='6.1.129-138.220.amzn2023.x86_64', version='#1 SMP PREEMPT_DYNAMIC Tue Feb 25 22:18:43 UTC 2025', machine='x86_64')


$ nvidia-smi
Wed Mar 26 21:26:31 2025
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 570.86.15              Driver Version: 570.86.15      CUDA Version: 12.8     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA H100 80GB HBM3          On  |   00000000:53:00.0 Off |                    0 |
| N/A   23C    P0            111W /  700W |     550MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA H100 80GB HBM3          On  |   00000000:64:00.0 Off |                    0 |
| N/A   24C    P0            112W /  700W |     538MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   2  NVIDIA H100 80GB HBM3          On  |   00000000:75:00.0 Off |                    0 |
| N/A   22C    P0            109W /  700W |     538MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   3  NVIDIA H100 80GB HBM3          On  |   00000000:86:00.0 Off |                    0 |
| N/A   27C    P0            115W /  700W |     538MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   4  NVIDIA H100 80GB HBM3          On  |   00000000:97:00.0 Off |                    0 |
| N/A   24C    P0            111W /  700W |     538MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   5  NVIDIA H100 80GB HBM3          On  |   00000000:A8:00.0 Off |                    0 |
| N/A   23C    P0            111W /  700W |     538MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   6  NVIDIA H100 80GB HBM3          On  |   00000000:B9:00.0 Off |                    0 |
| N/A   25C    P0            115W /  700W |     538MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   7  NVIDIA H100 80GB HBM3          On  |   00000000:CA:00.0 Off |                    0 |
| N/A   23C    P0            112W /  700W |     538MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A          435391      C   python                                  540MiB |
|    1   N/A  N/A          435391      C   python                                  528MiB |
|    2   N/A  N/A          435391      C   python                                  528MiB |
|    3   N/A  N/A          435391      C   python                                  528MiB |
|    4   N/A  N/A          435391      C   python                                  528MiB |
|    5   N/A  N/A          435391      C   python                                  528MiB |
|    6   N/A  N/A          435391      C   python                                  528MiB |
|    7   N/A  N/A          435391      C   python                                  528MiB |
+-----------------------------------------------------------------------------------------+
```

**Comment by jfc4050** [comment]
found this https://docs.jax.dev/en/latest/_autosummary/jax.lax.dot_general.html#jax.lax.dot_general

> preferred_element_type (DTypeLike | None | None) – Optional. This parameter controls the data type output by the dot product. By default, the output element type of this operation will match the lhs and rhs input element types under the usual type promotion rules. Setting preferred_element_type to a specific dtype will mean that the operation returns that element type. When precision is not a [DotAlgorithm](https://docs.jax.dev/en/latest/jax.lax.html#jax.lax.DotAlgorithm) or [DotAlgorithmPreset](https://docs.jax.dev/en/latest/jax.lax.html#jax.lax.DotAlgorithmPreset), preferred_element_type provides a hint to the compiler to accumulate the dot product using this data type.

so maybe what we have to do is to add a new primitive w/ the following 3 `preferred_element_type`s
* half precision for fwd
* half precision for dgrad
* full precision for wgrad

wondering a) is this the most idiomatic way to do this in JAX, and b) what to do for the non-GEMM parameters

**Comment by jfc4050** [comment]
^ did a quick prototype and this seems to work. would still be super interested if there are simpler/better ways and/or what should be done about non-GEMM parameters.
